File: 장세훈/LeetCode/medium/LetterCombinationsOfAPhoneNumber.java
class LetterCombinationsOfAPhoneNumber {
    /*
        https://leetcode.com/problems/letter-combinations-of-a-phone-number/description/
        17. Letter Combinations of a Phone Number

        주어진 숫자로 만들 수 있는 문자 조합을 구하는 문제.
        숫자별 문자들은 map으로 저장.
        bfs 알고리즘으로(재귀) 문자 조합을 만들어 answer에 저장
     */

    Map<Character, List<Character>> map = new HashMap<>();
    List<String> answer = new ArrayList<>();

    void init() {
        map.put('2', List.of('a', 'b', 'c'));
        map.put('3', List.of('d', 'e', 'f'));
        map.put('4', List.of('g', 'h', 'i'));
        map.put('5', List.of('j', 'k', 'l'));
        map.put('6', List.of('m', 'n', 'o'));
        map.put('7', List.of('p', 'q', 'r', 's'));
        map.put('8', List.of('t', 'u', 'v'));
        map.put('9', List.of('w', 'x', 'y', 'z'));

    }

    public List<String> letterCombinations(String digits) {
        if(digits.isEmpty()) return answer;
        init();
        dfs(0, digits, "");
        return answer;
    }

    void dfs(int i, String digits, String str) {
        if (i >= digits.length())
            answer.add(str);
        else {
            List<Character> characters = map.get(digits.charAt(i));
            for (Character character : characters) {
                dfs(i + 1, digits, str + character);
            }
        }
    }
}

File: 정수빈/LeetCode/medium/letter-combinations-of-a-phone-number.py
# https://leetcode.com/problems/letter-combinations-of-a-phone-number
# 설명: 백트레킹 문제인데 전혀 백트레킹처럼 안풀었다는... 숫자를 영문으로 변경하고 그에 맞는 모든 조합을 찾는 문제
# 난이도: medium
# 시간: 16분

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        alpha = {
            "2": list("abc"),
            "3": list("def"),
            "4": list("ghi"),
            "5": list("jkl"),
            "6": list("mno"),
            "7": list("pqrs"),
            "8": list("tuv"),
            "9": list("wxyz")
        }
        if len(digits) == 0:
            return []
        elif len(digits) == 1:
            return list(alpha[digits])

        elif len(digits) == 2:
            ans = []
            for d1 in alpha[digits[0]]:
                for d2 in  alpha[digits[1]]:
                    c = d1+d2
                    ans.append(c)
            return list(set(ans))
        elif len(digits) == 3:
            ans = []
            for d1 in alpha[digits[0]]:
                for d2 in  alpha[digits[1]]:
                    for d3 in  alpha[digits[2]]:
                        c = d1+d2+d3
                        ans.append(c)
            return list(set(ans))
        elif len(digits) == 4:
            ans = []
            for d1 in alpha[digits[0]]:
                for d2 in  alpha[digits[1]]:
                    for d3 in  alpha[digits[2]]:
                        for d4 in  alpha[digits[3]]:
                            c = d1+d2+d3+d4
                            ans.append(c)
            return list(set(ans))        

